apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/environment/PortalMetaServerProvider.java

```diff
@@ -0,0 +1,131 @@
+package com.ctrip.framework.apollo.portal.environment;
+
+import com.ctrip.framework.apollo.core.utils.ResourceUtils;
+import com.ctrip.framework.apollo.portal.util.KeyValueUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Only use in apollo-portal
+ * load all meta server address from
+ *  - System Property           [key ends with "_meta"]
+ *  - OS environment variable   [key ends with "_meta"]
+ *  - user's configuration file [key ends with ".meta"]
+ * when apollo-portal start up.
+ * @see com.ctrip.framework.apollo.core.internals.LegacyMetaServerProvider
+ * @author wxq
+ */
+public class PortalMetaServerProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(PortalMetaServerProvider.class);
+
+    /**
+     * environments and their meta server address
+     * properties file path
+     */
+    private static final String APOLLO_ENV_PROPERTIES_FILE_PATH = "apollo-env.properties";
+
+    // thread safe
+    private static volatile Map<Env, String> domains = initializeDomains();
+
+    public PortalMetaServerProvider() {
+
+    }
+
+    /**
+     * load all environment's meta address dynamically when this class loaded by JVM
+     */
+    private static Map<Env, String> initializeDomains() {
+        // find key-value from System Property which key ends with "_meta"
+        Map<String, String> metaServerAddressesFromSystemProperty = KeyValueUtils.filterWithKeyEndswith(System.getProperties(), "_meta");
+        // remove key's suffix "_meta"
+        metaServerAddressesFromSystemProperty = KeyValueUtils.removeKeySuffix(metaServerAddressesFromSystemProperty, "_meta".length());
+
+        // find key-value from OS environment variable which key ends with "_meta"
+        Map<String, String> metaServerAddressesFromOSEnvironment = KeyValueUtils.filterWithKeyEndswith(System.getenv(), "_meta");
+        // remove key's suffix "_meta"
+        metaServerAddressesFromOSEnvironment = KeyValueUtils.removeKeySuffix(metaServerAddressesFromOSEnvironment, "_meta".length());
+
+        // find key-value from properties file which key ends with ".meta"
+        Properties properties = new Properties();
```

Comment:
I think it might be better to also support setting meta server address from PortalDB.ServerConfig, which is the same place to store `apollo.portal.envs` and is therefore more intuitive, e.g. user can set env and env meta server address in `System Configuration` page. How do you think?

Reply:
You mean the meta server addresses are save in `apollo.portal.envs` which type is a `map`?
the config like follow:
```yml
apollo:
  portal:
    envs:
      dev: http://ip:port
      other: http://ip:port
      self-define: http://ip:port
```
And in `PortalDB.ServerConfig`'s `apollo.portal.envs`, there is a json object
```
{
      "dev"" : "http://ip:port"
      "other" : "http://ip:port"
      "self-define" : "http://ip:port"
}
```
So then portal conbine those config to get the meta server addresses?

Reply:
I think this feature (add environment in portal when running) will consume more code to implement.
How about when this feature (add environment in portal when start up) finished, next plan to do it?

Reply:
I think we can keep it as `add environment in portal when start up`, but we support configuring it via `System Configuration` page, e.g. we add a new configuration item like `apollo.portal.meta.servers` to store the mapping between env and meta server address.

We will still keep the `apollo.portal.envs` as it indicates the active environments so that there may be 3 environment mappings in `apollo.portal.meta.servers` but only 2 active envs in `apollo.portal.envs`.

If this involves much more effort, then it's okay to leave it when we support `add environment in portal when running`.

Reply:
Get it. 
It involves the Database operations. Add new key `apollo.portal.meta.servers`, and refresh it in runtime.
I feel that the database operation is a bit difficult.
I don't know how to add this feature now.